File: dp/dp_g.py
"""
AtCoder Educational DP Contest
https://atcoder.jp/contests/dp/tasks/dp_g

G - Longest path

"""
import collections
import functools
import sys
sys.setrecursionlimit(200000)


def get_adj_list(pairs):
    adj = collections.defaultdict(list)
    for (i, j) in pairs:
        adj[i].append(j)
    return adj


def solve(pairs):
    adj_list = get_adj_list(pairs)

    @functools.lru_cache(200000)
    def dfs(u):
        res = -1
        for v in adj_list[u]:
            res = max(res, dfs(v))
        return 1 + res

    depths = (dfs(u) for u in range(1, N+1))
    return max(depths, default=0)


N, M = map(int, input().split())
pairs = [[int(s) for s in input().split()] for _ in range(M)]
res = solve(pairs)
print(res)

File: dp/dp_g.cc
// https://atcoder.jp/contests/dp/tasks/dp_g
#include <iostream>
#include <map>
#include <stack>
#include <vector>
using namespace std;

int N, M;
vector<vector<int>> adjlist(100004);
vector<int> memo(100004);

int depth(int start)
{
    if (memo[start] > 0)
        return memo[start];

    int res = 0;
    for (auto& next : adjlist[start]) {
        res = max(res, 1 + depth(next));
    }
    return memo[start] = res;
}

int main()
{
    cin >> N >> M;
    for (int i = 0; i < M; ++i) {
        int tmpx, tmpy;
        cin >> tmpx >> tmpy;
        adjlist[tmpx].push_back(tmpy);
    }

    int res = 0;
    for (int i = 1; i <= N; ++i)
        res = max(res, depth(i));

    cout << res << endl;

    return 0;
}